beacon-chain/sync/initial-sync/blocks_fetcher.go

```diff
@@ -317,30 +325,117 @@ func (f *blocksFetcher) handleRequest(ctx context.Context, start primitives.Slot
 	if f.mode == modeStopOnFinalizedEpoch {
 		highestFinalizedSlot := params.BeaconConfig().SlotsPerEpoch.Mul(uint64(targetEpoch + 1))
 		if start > highestFinalizedSlot {
-			response.err = fmt.Errorf("%w, slot: %d, highest finalized slot: %d",
-				errSlotIsTooHigh, start, highestFinalizedSlot)
+			response.err = fmt.Errorf(
+				"%w, slot: %d, highest finalized slot: %d",
+				errSlotIsTooHigh, start, highestFinalizedSlot,
+			)
+
 			return response
 		}
 	}
 
 	response.bwb, response.blocksFrom, response.err = f.fetchBlocksFromPeer(ctx, start, count, peers)
 	if response.err == nil {
-		pid, bwb, err := f.fetchBlobsFromPeer(ctx, response.bwb, response.blocksFrom, peers)
+		pid, err := f.fetchSidecars(ctx, response.blocksFrom, peers, response.bwb)
 		if err != nil {
+			log.WithError(err).Error("Failed to fetch sidecars")
 			response.err = err
 		}
-		response.bwb = bwb
+
 		response.blobsFrom = pid
 	}
+
 	return response
 }
 
-// fetchBlocksFromPeer fetches blocks from a single randomly selected peer.
+// fetchSidecars fetches sidecars corresponding to blocks in `response.bwb`.
+// It mutates `Blobs` and `Columns` fields of `response.bwb` with fetched sidecars.
+// `pid` is the initial peer to request blob from (usually the peer from which the block originated),
+// `peers` is a list of peers to use for the request blobs if `pid` fails.
+// It returns the peer ID from which blobs were fetched (if any).
+func (f *blocksFetcher) fetchSidecars(ctx context.Context, pid peer.ID, peers []peer.ID, bwScs []blocks.BlockWithROSidecars) (peer.ID, error) {
+	samplesPerSlot := params.BeaconConfig().SamplesPerSlot
+
+	// Find the first block with a slot greater than or equal to the first Fulu slot.
+	// (Blocks are sorted by slot.)
+	firstFuluIndex := sort.Search(len(bwScs), func(i int) bool {
+		return bwScs[i].Block.Version() >= version.Fulu
+	})
+
+	blocksWithBlobs := bwScs[:firstFuluIndex]
+	blocksWithDataColumns := bwScs[firstFuluIndex:]
+
+	if len(blocksWithBlobs) == 0 && len(blocksWithDataColumns) == 0 {
```

Comment:
Isn't this equivalent to 
```suggestion
	if len(bwScs) == 0 {
```
Which can be checked before the search above

Reply:
Yes. I don't accept your suggestion as is since this change can be done earlier in the function.
Fixed in 901f6b6e6cac93999a2276a87a8145698f48c90e.